File: misc/dist-elijahs_sequential_ctf/elij.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n; cin >> n;
    vector<int> a(n);
    for (int &x : a) cin >> x;

    const int64_t NEG_INF = -1e15;
    vector<int64_t> dp(3, NEG_INF);
    dp[0] = dp[1] = dp[2] = NEG_INF;

    // Satisfaction lookup
    int s[3][3] = {
        {0, 2, 3},
        {4, 0, 5},
        {1, 6, 0}
    };

    for (int c : a) {
        vector<int64_t> new_dp = dp;
        if (new_dp[c] < 0) new_dp[c] = 0; // start subsequence here

        for (int prev = 0; prev < 3; prev++) {
            if (dp[prev] == NEG_INF) continue;
            int64_t val = dp[prev] + s[prev][c];
            if (val > new_dp[c]) new_dp[c] = val;
        }

        dp = move(new_dp);
    }

    cout << max({dp[0], dp[1], dp[2]}) << "\n";
    return 0;
}

File: misc/dist-elijahs_sequential_ctf/elij.py
import sys
input = sys.stdin.readline

def main():
    n = int(input())
    categories = list(map(int, input().split()))
    
    satisfaction = {
        (0,1): 2,
        (1,2): 5,
        (0,2): 3,
        (1,0): 4,
        (2,0): 1,
        (2,1): 6
    }
    
    NEG_INF = -10**15
    dp = [NEG_INF]*3
    
    get_satis = satisfaction.get  # local ref to speed up
    
    for c in categories:
        new_dp = dp[:]
        if new_dp[c] < 0:
            new_dp[c] = 0  # start subsequence here if better
        
        for prev in range(3):
            prev_val = dp[prev]
            if prev_val != NEG_INF:
                val = prev_val + get_satis((prev,c), 0)
                if val > new_dp[c]:
                    new_dp[c] = val
        
        dp = new_dp
    
    print(max(dp))

if __name__ == "__main__":
    main()
